### ISSUE TITLE
[Feature Request] PEP 526 support (also known as: "Statically type-check variable annotations, yo!")

### ISSUE THREAD

**Issue by leycec** [beartype/beartype#105]
<sup>[Shameless copy-paste job courtesy @langfield.](https://github.com/beartype/beartype/issues/94)</sup>

For example, do we ever intend it to detect the issue in either of these:

```python
from beartype import beartype

@beartype
def fn(n: int) -> None:
    n = "waffle"

@beartype
def fn2() -> None: 
    n = 5
    n = "waffle"
```

Or is that out of the scope of the project?

Basically, I would like to be able to completely ditch mypy in favor of this project, and this is what's holding me back.

**Comment by leycec** [comment]
This question can be succinctly rephrased as follows:

> Would we like to destroy mypy and every other static type checker?

The answer to this and many other questions is:

> **...yes.** Why, yes. We would like to do that thing.

We sympathesize with your urge to dump mypy into a ditch while wearing vinyl gloves which you then quietly incinerate. Dumping mypy into a ditch is entirely **in-scope** for `beartype`. That's why we're here, after all – to realize that utopian dream we all share.

Basically, we regard:

* Static type-checking as the ignoble past we all put up with because we have to.
* Runtime type-checking as the Halcyon Golden Age we are rocketing towards at risky speeds that may not be sustainable, safe, or a good idea. *But it sure is fun.*

## So How We Do This, Bro?

We go [full AST](https://docs.python.org/3/library/ast.html) on this mutha. I even have internal `#FIXME:` comments hidden somewhere within the `beartype` codebase documenting a probable route to pragmatically achieving this.

Of course, ASTs are a fathomless black hole from which few return alive and even fewer return sane. This may not necessarily happen within the expected lifetime of my life is what I'm saying. A drip-feed of corporate and/or governmental sponsorship would probably get us there sooner than later. Until Money Shangri La materializes, however, this is lower on the priority list than you'd like. <sup>*Ugh!*</sup>

## tl;dr

**Money.** That's what can make this happen. :money_mouth_face:

**Comment by leycec** [comment]
**\o/**

Fellow GitHubber @awf has now successfully implemented the first practical example I've seen of performing static type-checking at runtime via [the new third-party `@awfutils.typecheck` decorator](https://github.com/awf/awfutils#typecheck). Since that decorator only currently supports `ininstance()`-able classes rather than PEP-compliant type hints, it isn't *quite* ready for full-throttle inclusion in `@beartype`. Still, everything works shockingly well for simple type hints that are classes:

```python
@typecheck
def foo(x : int, y : float):
  z : int = x * y # Now it raises AssertionError: z not int 
  w : float = z * 3.2
  return w

foo(3, 1.3) # Error comes from this call
```

The core implementation has phenomenal merit. `@awfutils.typecheck` could eventually serve as the basis for a PEP-compliant solution in `@beartype` itself. Thunderous kudos to @awf for rolling out such impressive tech. This Ontario thundercloud is for you, @awf. :cloud_with_lightning:

**Comment by leycec** [comment]
**Arise, dark necrobump!** Let's update this after a lifetime of neglecting this. `\o/`

It turns out that *this* feature request implicitly requires *another* feature request: #43 (i.e., "Automagical Import Hook") as a mandatory dependency. First, we do #43; then, we ~~take Berlin~~ do this. Let's dig a bit deeper into exactly what I am talking about here.

The `beartype.claw` submodule *is* an unfinished public API that mostly currently exists and intends to do exactly what everyone wants in both this feature request *and* #43: e.g.,

```python
# In your package-level `torch.onnx.__init__` submodule:
from beartype.claw import beartype_package
beartype_package()  # <-- *BOOM*!
```

As you correctly suspect, `beartype.claw.beartype_package()` is @beartype's **import hook.** Pass that function either:

* The fully-qualified name of any package or module that has yet to be fully imported (e.g., `"torch.onnx"`).
* Nothing, in which case that defaults to the name of the parent package of the current module (e.g., `"torch.onnx"` in the above example).

That function then registers a new import hook with Python's low-level `importlib` machinery implicitly decorating *all* classes and functions (that have yet to be imported from the passed package or module) with `@beartype`. In short, it just works. Except it doesn't – yet.

## I Can't Believe @leycec Is Still Typing

But wait... *there's more!* Everything above is pretty exciting, because it will save everyone precious time, attention, and life force. But there's something even more exciting. `</gasp>`

`beartype.claw.beartype_package()` isn't just syntactic sugar for convenience; calling that function will also enable **hybrid static-runtime type-checking.** Once you call that function, `@beartype` will begin statically analysing your code in a similar manner to that of traditional static type-checkers like mypy and `pyright`. This will make `@beartype` the first hybrid static-runtime type-checker. Flex those burly biceps, @beartype. :muscle: 

Specifically, after calling `beartype.claw.beartype_package()`, @beartype will begin type-checking **annotated assignments** at any scope (including both global scope and local scope embedded in a callable). For example:

```python
# In *ANY* submodule of the `torch.onnx` package:
beartype_will_typecheck_this_and_raise_an_exception: int = 'Not an int.'

def muh_func() -> None:
    beartype_will_typecheck_this_too_and_silently_accept_it: str = 'I am string!'
```

All of that's implemented – but none of that's tested. It almost certainly does *not* work yet. Sadly, something always explodes on our issue tracker before I can get back to finalizing the `beartype.claw` API. My face is exhaling. :face_exhaling: 

I acknowledge that this is critically important functionality. I also acknowledge that I am lazy and would prefer to continue reading the [twelve-volume *Classroom of the Elite* light novel](https://myanimelist.net/manga/89357/Youkoso_Jitsuryoku_Shijou_Shugi_no_Kyoushitsu_e) that I've been reading for literally months. I'm emotionally invested in the escapades of high-IQ high school students from a dystopian techno-futuresque Japan. How did this happen to me!?!?

**Comment by leycec** [comment]
*...lol.* Just realized we resolved this a few months ago with @beartype 0.15.0. For everyone playing along at home, <sup>@beartype: *It's Fun, Kids!*</sup> please see [my exhausting release notes for @beartype 0.15.0 for further details](https://github.com/beartype/beartype/discussions/257). For everyone who doesn't want to read a tedious dissertation on the nature of import hooks, this is your **tl;dr:**

```python
# In root "{your_package}.__init__" submodule:
from beartype.claw import beartype_this_package
beartype_this_package()
```

**That's it.** Congrats. @beartype will now statically type-check *all* PEP 526-compliant variable annotations across your entire codebase. When you play with @beartype, you're playing with power. :muscle: :bear:
